File: quick_sort.js
function quickSort(arr) {
    // Base case: arrays with 0 or 1 element are already sorted
    if (arr.length <= 1) {
        return arr
    }

    const pivot = arr[arr.length - 1]
    const left = [] // Elements smaller than or equal to the pivot
    const right = [] // Elements greater than the pivot

    // Partition the array into subarrays
    for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i] <= pivot) {
            left.push(arr[i])
        } else {
            right.push(arr[i])
        }
    }

    // Recursively sort the subarrays and combine the result
    return [...quickSort(left), pivot, ...quickSort(right)]
}

// Test the algorithm
const unsortedArray = [10, 15, 23, 25, 22, 69, 80, 100, 45, 85, 38]
console.log('Unsorted array', unsortedArray)

const sortedArray = quickSort(unsortedArray)
console.log('Sorted array', sortedArray)

File: quick_sort.py
def quick_sort(arr):
    if len(arr) <= 1:
        return arr  # Base case: arrays with 1 or no elements are already sorted

    # Choose the pivot (last element in this case)
    pivot = arr[-1]
    left = [x for x in arr[:-1] if x <= pivot]  # Elements smaller than or equal to the pivot
    right = [x for x in arr[:-1] if x > pivot]  # Elements greater than the pivot

    # Recursion on the left and right subarrays
    return quick_sort(left) + [pivot] + quick_sort(right)

# Test
unsorted_array = [10, 80, 30, 90, 40, 50, 70]
sorted_array = quick_sort(unsorted_array)
print("Unsorted Array:", unsorted_array)
print("Sorted Array:", sorted_array)
